src/ai/backend/manager/models/etcd.py

```diff
@@ -0,0 +1,257 @@
+    )
+    async def mutate(
+        cls, root, info: graphene.ResolveInfo, hostname: str, props: CreateContainerRegistryInput
+    ) -> CreateContainerRegistry:
+        ctx: GraphQueryContext = info.context
+        if await ctx.shared_config.etcd.get(f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}") is None:
+            raise ObjectNotFound(object_name=f"registry: {hostname}")
+
+        raw_config = {
+            "": props.url,
+            "type": props.container_type,
+            "project": props.project,
+            "username": props.username,
+            "password": props.password,
+        }
+        container_registry_iv.check(raw_config)
+        config = {key: value for key, value in raw_config.items() if value is not None}
+
+        log.info(
+            "ETCD.MODIFY_CONTAINER_REGISTRY (ak:{}, key: {}, hostname:{}, config:{})",
+            ctx.access_key,
+            ETCD_CONTAINER_REGISTRY_KEY,
+            hostname,
+            config,
+        )
+        updates = {}
+
+        def flatten(o):
+            for k, v in o.items():
+                if k == "":
+                    inner_prefix = f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}"
+                else:
+                    inner_prefix = f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}/{k}"
+                if isinstance(v, Mapping):
+                    flatten(v)
+                else:
+                    updates[inner_prefix] = v
+
+        flatten(config)
+        # TODO: chunk support if there are too many keys
+        if len(updates) > 16:
+            raise InvalidAPIParameters("Too large update! Split into smaller key-value pair sets.")
+        await ctx.shared_config.etcd.put_dict(updates)
+        return cls(result="ok")
+
+
+class DeleteContainerRegistry(graphene.Mutation):
+    allowed_roles = (UserRole.SUPERADMIN,)
+    result = graphene.String()
+
+    class Arguments:
+        hostname = graphene.String(required=True)
+
+    @classmethod
+    @privileged_mutation(
+        UserRole.SUPERADMIN,
+        lambda id, **kwargs: (None, id),
+    )
+    async def mutate(
+        cls,
+        root,
+        info: graphene.ResolveInfo,
+        hostname: str,
+    ) -> DeleteContainerRegistry:
+        ctx: GraphQueryContext = info.context
+        raw_registries = await ctx.shared_config.etcd.get_prefix_dict(ETCD_CONTAINER_REGISTRY_KEY)
+        registries = t.Mapping(t.String, container_registry_iv).check(raw_registries)
+        try:
+            del registries[hostname]
+        except KeyError:
+            raise ObjectNotFound(object_name=f"registry: {hostname}")
+        log.info(
+            "ETCD.DELETE_CONTAINER_REGISTRY (ak:{}, key: {}, hostname:{})",
+            ctx.access_key,
+            ETCD_CONTAINER_REGISTRY_KEY,
+            hostname,
+        )
+        await ctx.shared_config.etcd.delete_prefix(f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}")
+        await ctx.shared_config.etcd.put_prefix(ETCD_CONTAINER_REGISTRY_KEY, registries)
```

Comment:
Let's think of any good way to reduce transactions!

Reply:
Currently there is simple way to group multiple mutation calls to `ctx.shared_config.etcd` unless we break into our thin `ai.backend.common.etcd` abstraction layer.

As mutation of container registries rarely happen in production, I think this is fine for now. We could introduce context-based transaction support in [etcd-client-py](https://github.com/lablup/etcd-client-py) and the future `ai.backend.common.etcd` API.